src/Microsoft.Graph.Cli.Core/Authentication/EnvironmentCredential.cs

```diff
@@ -0,0 +1,132 @@
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Obtains a token from the Azure Active Directory service, using the specified client details specified in the environment variables
+        /// AZURE_TENANT_ID, AZURE_CLIENT_ID, and AZURE_CLIENT_SECRET or AZURE_USERNAME and AZURE_PASSWORD to authenticate.
+        /// Acquired tokens are cached by the credential instance. Token lifetime and refreshing is handled automatically. Where possible,
+        /// reuse credential instances to optimize cache effectiveness.
+        /// </summary>
+        /// <remarks>
+        /// If the environment variables AZURE_TENANT_ID, AZURE_CLIENT_ID, and AZURE_CLIENT_SECRET are not specified, the default <see cref="AccessToken"/>
+        /// </remarks>
+        /// <param name="requestContext">The details of the authentication request.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> controlling the request lifetime.</param>
+        /// <returns>An <see cref="AccessToken"/> which can be used to authenticate service client calls.</returns>
+        public override AccessToken GetToken(TokenRequestContext requestContext, CancellationToken cancellationToken = default)
+        {
+            return GetTokenImplAsync(false, requestContext, cancellationToken).GetAwaiter().GetResult();
+        }
+
+        /// <summary>
+        /// Obtains a token from the Azure Active Directory service, using the specified client details specified in the environment variables
+        /// AZURE_TENANT_ID, AZURE_CLIENT_ID, and AZURE_CLIENT_SECRET or AZURE_USERNAME and AZURE_PASSWORD to authenticate.
+        /// Acquired tokens are cached by the credential instance. Token lifetime and refreshing is handled automatically. Where possible,
+        /// reuse credential instances to optimize cache effectiveness.
+        /// </summary>
+        /// <remarks>
+        /// If the environment variables AZURE_TENANT_ID, AZURE_CLIENT_ID, and AZURE_CLIENT_SECRET are not specified, the default <see cref="AccessToken"/>
+        /// </remarks>
+        /// <param name="requestContext">The details of the authentication request.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> controlling the request lifetime.</param>
+        /// <returns>An <see cref="AccessToken"/> which can be used to authenticate service client calls, or a default <see cref="AccessToken"/>.</returns>
+        public override async ValueTask<AccessToken> GetTokenAsync(TokenRequestContext requestContext, CancellationToken cancellationToken = default)
+        {
+            return await GetTokenImplAsync(true, requestContext, cancellationToken).ConfigureAwait(false);
+        }
+
+        private async ValueTask<AccessToken> GetTokenImplAsync(bool async, TokenRequestContext requestContext, CancellationToken cancellationToken)
```

Comment:
That's the same implementation that Azure identity have. I was trying to keep the API as close as possible to the original. I'll remove it for now since I don't use it...

Reply:
Actually, on second thought, The GetToken function is part of the contract for a token credential. I can't remove it.

Reply:
I believe @baywet is referring to the Boolean flag used by the private method `GetTokenImplAsync`, the private method does not need to call `Credential.GetToken` but just call `Credential.GetTokenAsync`, the `GetToken` overrides do not need to pass the Boolean parameter as the private method is already called synchronously by using `GetAwaiter.GetResult`

Reply:
The boolean is to select which method to call on the base credential. This matches Azure Identity's design.

Reply:
using a flag to drive the async behaviour feels super hacky. I'd be curious to know what's the rationale behind it when the tasking system already supports running tasks synchronously. They might have done that because some early winRT implementations of the Task Scheduler could end-up in a deadlock with the UI thread back in the days. But that's a problem we don't have.

Reply:
Ok, what I'll do is remove the shared GetTokenImplAsync so we don't have the issue.